model/histogram/float_histogram.go

```diff
@@ -115,6 +115,295 @@ func (h *FloatHistogram) ZeroBucket() FloatBucket {
 	}
 }
 
+// Scale scales the FloatHistogram by the provided factor, i.e. it scales all
+// bucket counts including the zero bucket and the count and the sum of
+// observations. The bucket layout stays the same. This method changes the
+// receiving histogram directly (rather than acting on a copy). It returns a
+// pointer to the receiving histogram for convenience.
+func (h *FloatHistogram) Scale(factor float64) *FloatHistogram {
+	h.ZeroCount *= factor
+	h.Count *= factor
+	h.Sum *= factor
+	for i := range h.PositiveBuckets {
+		h.PositiveBuckets[i] *= factor
+	}
+	for i := range h.NegativeBuckets {
+		h.NegativeBuckets[i] *= factor
+	}
+	return h
+}
+
+// Add adds the provided other histogram to the receiving histogram. Count, Sum,
+// and buckets from the other histogram are added to the corresponding
+// components of the receiving histogram. Buckets in the other histogram that do
+// not exist in the receiving histogram are inserted into the latter. The
+// resulting histogram might have buckets with a population of zero or directly
+// adjacent spans (offset=0). To normalize those, call the Compact method.
+//
+// This method returns a pointer to the receiving histogram for convenience.
+//
+// IMPORTANT: This method requires the Schema and the ZeroThreshold to be the
+// same in both histograms. Otherwise, its behavior is undefined.
+// TODO(beorn7): Change that!
+func (h *FloatHistogram) Add(other *FloatHistogram) *FloatHistogram {
+	h.ZeroCount += other.ZeroCount
+	h.Count += other.Count
+	h.Sum += other.Sum
+
+	// TODO(beorn7): If needed, this can be optimized by inspecting the
+	// spans in other and create missing buckets in h in batches.
+	iSpan, iBucket := -1, -1
+	var iInSpan, index int32
+	for it := other.PositiveBucketIterator(); it.Next(); {
+		b := it.At()
+		h.PositiveSpans, h.PositiveBuckets, iSpan, iBucket, iInSpan = addBucket(
+			b, h.PositiveSpans, h.PositiveBuckets, iSpan, iBucket, iInSpan, index,
+		)
+		index = b.Index
+	}
+	iSpan, iBucket = -1, -1
+	for it := other.NegativeBucketIterator(); it.Next(); {
+		b := it.At()
+		h.NegativeSpans, h.NegativeBuckets, iSpan, iBucket, iInSpan = addBucket(
+			b, h.NegativeSpans, h.NegativeBuckets, iSpan, iBucket, iInSpan, index,
+		)
+		index = b.Index
+	}
+	return h
+}
+
+// Sub works like Add but subtracts the other histogram.
+//
+// IMPORTANT: This method requires the Schema and the ZeroThreshold to be the
+// same in both histograms. Otherwise, its behavior is undefined.
+// TODO(beorn7): Change that!
+func (h *FloatHistogram) Sub(other *FloatHistogram) *FloatHistogram {
+	h.ZeroCount -= other.ZeroCount
+	h.Count -= other.Count
+	h.Sum -= other.Sum
+
+	// TODO(beorn7): If needed, this can be optimized by inspecting the
+	// spans in other and create missing buckets in h in batches.
```

Comment:
I have a simpler solution for this using the `MergeBucketIterator` that I implemented (see [here](https://github.com/prometheus/prometheus/pull/9935/commits/7e151d92132255279cf98f915f03ab778961b809)).

We can independently merge the span and use `MergeBucketIterator` to naively populate the buckets and we can expect them to be in the order of the merged spans.

And to extend it for sum(), we can add a global `Add(hists ...FloatHistogram) FloatHistogram` function which merges the span across all the histograms and users `MergeBucketIterator` to get the buckets. It should be then easier to also extend it to support changing schema at a central place. I will take a dig at it with sum() implementation and we can use that here instead.

Reply:
Turns out it is easier than that. The buckets from merged iterator would also have the index for the bucket. So we can build the spans on the fly!

Reply:
This proposed way performs worse https://github.com/prometheus/prometheus/pull/9947